File: 2015/kotlin/src/main/kotlin/nl/sanderp/aoc/common/Math.kt
package nl.sanderp.aoc.common

import kotlin.math.PI
import kotlin.math.abs

/**
 * Calculates the greatest common divisor between two numbers.
 */
fun gcd(a: Int, b: Int): Int = if (b == 0) a else gcd(b, a % b)

/**
 * Calculates the least common multiple between two numbers.
 */
fun lcm(a: Int, b: Int): Int = abs(a * b) / gcd(a, b)

fun Double.radians() = PI * this / 180.0
fun Int.radians() = toDouble().radians()

File: 2020/fsharp/Common/Math.fs
namespace Common

module Math =
    open System

    /// Calculate the Greatest Common Divisor between a and b
    let rec gcd a b = if b = 0 then a else gcd b (a % b)

    /// Calculate the Least Common Multiple between a and b
    let lcm a b = abs (a * b) / gcd a b

    /// Convert degrees to radians
    let rad n = Math.PI * n / 180.0

    /// Rotate a point by degrees
    let rotate (x, y) deg =
        let angle = rad deg
        let x' = x * cos angle - y * sin angle
        let y' = y * cos angle + x * sin angle
        (x', y')
